### PR TITLE
🚀 Feature: Add idempotency middleware

### PR DESCRIPTION
**Pull request by leonklingele**
## Description

Add idempotency middleware.
Implements https://github.com/gofiber/fiber/issues/2163.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist:

- [x] For new functionalities I follow the inspiration of the express js framework and built them similar in usage
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation - https://github.com/gofiber/docs for https://docs.gofiber.io/
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] If new dependencies exist, I have checked that they are really necessary and agreed with the maintainers/community (we want to have as few dependencies as possible)
- [x] I tried to make my code as fast as possible with as few allocations as possible
- [ ] For new code I have written benchmarks so that they can be analyzed and improved

### CODE REVIEW

File: middleware/idempotency/storage.go
Comment (efectn) [inline middleware/idempotency/storage.go:8]: Why did you create new storage interface. I think you can make this like csrf, cache middlewares

Comment (leonklingele) [reply middleware/idempotency/storage.go:8]: We don't need the following `fiber.Storage` methods:

- `Delete(key string) error`
- `Reset() error`
- `Close() error`

All types satisfying the `idempotency.Storage` interface also implement the `fiber.Storage` interface.

Comment (efectn) [reply middleware/idempotency/storage.go:8]: Storage is generic interface for various fiber middlewares. I think we shouldnt create new interface. I can't see any desadvantage of using fiber.Storage. We should make code as simple as possible and we should avoid much copying i think.

Comment (GalvinGao) [reply middleware/idempotency/storage.go:8]: Would suggest the same here, and maybe also document it as something like:

> Those methods are unused so simply put a `panic("unexpected call")` would be sufficient.

**Review by GalvinGao — CHANGES_REQUESTED** [review]
I would argue the same as in the conversation https://github.com/gofiber/fiber/pull/2253/files#r1038919642 to have consistency with `fiber.Storage`. Otherwise wonderful work!

**Comment by efectn** [comment]
Why did you target `v3-beta` branch? I think this can be a part of v2.

Also apply PR template into your PRs please.

**Comment by leonklingele** [comment]
@efectn I'll backport this to v2 once merged.

File: middleware/idempotency/storage.go
Comment (leonklingele) [reply middleware/idempotency/storage.go]: Using a `fiber.Storage` now.

**Comment by leonklingele** [comment]
Soo, finally this is now ready for review. I removed the `Storage` interface, switched to `msgp` for marshalling, and improved some other things.

**Comment by leonklingele** [comment]
Tests seem to be flaky: https://github.com/gofiber/fiber/actions/runs/3615636468/jobs/6092857426
Is this a known issue?

**Comment by leonklingele** [comment]
Another, maybe different failure: https://github.com/gofiber/fiber/actions/runs/3615669563/jobs/6092912570

File: middleware/idempotency/idempotency.go
Comment (leonklingele) [inline middleware/idempotency/idempotency.go:116]: I wonder if we need to use `utils.CopyBytes` here.

**Comment by efectn** [comment]
Can you change branch? Then, i'll review it

**Comment by leonklingele** [comment]
Let's make this go in v3 first and I'll open up another PR targeting v2 later on.

**Comment by leonklingele** [comment]
Anyone gotten the chance to review this already? I'm working on a project which might benefit from having this.

**Comment by ReneWerner87** [comment]
Will do it tomorrow

**Review by ReneWerner87 — APPROVED** [review]
implementation looks very good

a few small comments/improvements

- benchmark for the middleware usage would be good, so we can improve it later on
- instead of the clousure i would maybe try to use a function that doesn't need to be created on-the-fly every time
- with the locals thing i'm not quite sure, what happens e.g. if you define the idempotency middleware multiple times, normally there shouldn't be any collisions as they should be used on different routes then ... currently quite not sure with that
- in some places we can optimize and save allocations, like in the ToLower( there exists a fn which does not cause allocations) and other places where only comparisons are done
-> for comparisons no copies need to be made, only if the data should be kept

**Comment by ReneWerner87** [comment]
@leonklingele will wait for the others for the feedback and then i merge it

@efectn @GalvinGao do you have any other points, otherwise i would merge the pull request

**Comment by leonklingele** [comment]
Thanks for the review! I’ll push the updates later today.

Which calls can be optimized? I’ve only found the single call to strings.ToLower used when doing the map lookup which we could replace by utils.ToLower

Also, did you see my question regarding using utils.CopyBytes when storing the request body? I think we need to use it.

**Comment by ReneWerner87** [comment]
Yes use the copy there, i was not sure about this

**Review by GalvinGao — APPROVED** [review]
lgtm!

**Comment by leonklingele** [comment]
I've rebased & addressed @ReneWerner87's points. ptal

**Review by efectn — APPROVED** [review]

**Review by ReneWerner87 — APPROVED** [review]

**Pull request merged**
